File: bindings/clojure/src/rockbox/ffi.clj
(ns rockbox.ffi
  "Runtime loader for the prebuilt librockbox_ffi shared library, built on the
  Java Foreign Function & Memory API (JEP 454, stable since JDK 22).

  Nothing is linked at build time: the library is located at runtime and every
  function is bound to a MethodHandle downcall. This mirrors
  include/rockbox_ffi.h — keep the descriptors in sync with the C ABI."
  (:require [clojure.string :as str]
            [clojure.java.io :as io])
  (:import [java.lang.foreign Linker Linker$Option FunctionDescriptor ValueLayout
            MemoryLayout Arena MemorySegment SymbolLookup]
           [java.lang.invoke MethodHandle]
           [java.io File]))

;; Layout shorthands (LP64: size_t / uint64_t / int64_t are all 8 bytes).
(def PTR ValueLayout/ADDRESS)   ; opaque pointer
(def I32 ValueLayout/JAVA_INT)  ; int32_t / uint32_t
(def I64 ValueLayout/JAVA_LONG) ; int64_t / uint64_t / size_t
(def F32 ValueLayout/JAVA_FLOAT)
(def I16 ValueLayout/JAVA_SHORT)
(def BOOL ValueLayout/JAVA_BOOLEAN)

;; ---- library location -------------------------------------------------

(defn- native-target
  "The `<os>-<arch>` release target for the running JVM, or nil if unsupported."
  []
  (let [os (str/lower-case (System/getProperty "os.name"))
        arch (str/lower-case (System/getProperty "os.arch"))
        o (cond
            (or (str/includes? os "mac") (str/includes? os "darwin")) "darwin"
            (str/includes? os "linux") "linux"
            (str/includes? os "freebsd") "freebsd"
            (str/includes? os "netbsd") "netbsd")
        a (cond
            (#{"aarch64" "arm64"} arch) "arm64"
            (#{"amd64" "x86_64" "x64"} arch) "x64")]
    (when (and o a) (str o "-" a))))

(defn- extract-bundled
  "Extract the platform's bundled native/<target>/librockbox_ffi.<ext> classpath
  resource to a temp file and return its path, or nil if none is bundled.
  `record` logs the attempted resource for error reporting."
  [record]
  (when-let [target (native-target)]
    (let [ext (if (str/starts-with? target "darwin") "dylib" "so")
          resource (str "native/" target "/librockbox_ffi." ext)]
      (record (str "classpath:" resource))
      (when-let [stream (.getResourceAsStream (clojure.lang.RT/baseLoader) resource)]
        (with-open [in stream]
          (let [tmp (File/createTempFile "librockbox_ffi" (str "." ext))]
            (.deleteOnExit tmp)
            (io/copy in tmp)
            (.getAbsolutePath tmp)))))))

(defn- locate-lib
  "Locate the shared library. Precedence:
     1. ROCKBOX_FFI_LIB env var (explicit override)
     2. the native lib bundled in the jar for this OS/arch (published pkg)
     3. target/release, walking up from the working directory (repo checkout)"
  ^String []
  (let [names ["librockbox_ffi.dylib" "librockbox_ffi.so" "rockbox_ffi.dll"]
        tried (java.util.ArrayList.)
        attempt (fn [^String path]
                  (.add tried path)
                  (when (.exists (File. path)) path))]
    (or
     (when-let [env (System/getenv "ROCKBOX_FFI_LIB")] (attempt env))
     (extract-bundled (fn [s] (.add tried s)))
     (loop [dir (.getAbsoluteFile (File. (System/getProperty "user.dir")))]
       (when dir
         (or (some (fn [sub] (some #(attempt (.getPath (File. dir (str sub "/" %)))) names))
                   ["target/release" "Libs"])
             (recur (.getParentFile dir)))))
     (throw (ex-info (str "could not locate librockbox_ffi shared library. Set "
                          "ROCKBOX_FFI_LIB or run `cargo build --release -p "
                          "rockbox-ffi`. Tried:\n  " (str/join "\n  " tried))
                     {:tried (vec tried)})))))

;; Lazy: even Linker/nativeLinker is a Foreign-API call, kept out of ns load so
;; requiring this namespace does no FFM work (cljdoc analysis, older JDKs).
(def ^:private linker (delay (Linker/nativeLinker)))

;; The library outlives every call, so it lives in the global arena (never
;; unloaded). Loaded lazily on first FFI use — merely `require`-ing this
;; namespace (e.g. for cljdoc doc analysis) must NOT touch native code.
(def ^:private lookup (delay (SymbolLookup/libraryLookup (locate-lib) (Arena/global))))

(defn- fd-of ^FunctionDescriptor [ret args]
  (FunctionDescriptor/of ret (into-array MemoryLayout args)))

(defn- fd-void ^FunctionDescriptor [args]
  (FunctionDescriptor/ofVoid (into-array MemoryLayout args)))

(defn- dc ^MethodHandle [^String name ^FunctionDescriptor fd]
  (let [sym (-> @lookup (.find name)
                (.orElseThrow (reify java.util.function.Supplier
                                (get [_] (ex-info (str "symbol " name " not found in librockbox_ffi") {})))))]
    (.downcallHandle @linker sym fd (make-array Linker$Option 0))))

;; Bound lazily (a delay): the downcall handles are built on first FFI call,
;; not at namespace load, so requiring this ns never dlopens the library.
(def ^:private handles
  (delay
   {;; ---- deallocators / version ----
   :abi-version                (dc "rb_ffi_abi_version" (fd-of I32 []))
   :string-free                (dc "rb_string_free" (fd-void [PTR]))
   :buffer-free                (dc "rb_buffer_free" (fd-void [PTR I64]))
   ;; ---- DSP ----
   :dsp-new                    (dc "rb_dsp_new" (fd-of PTR [I32]))
   :dsp-free                   (dc "rb_dsp_free" (fd-void [PTR]))
   :dsp-set-input-frequency    (dc "rb_dsp_set_input_frequency" (fd-void [PTR I32]))
   :dsp-flush                  (dc "rb_dsp_flush" (fd-void [PTR]))
   :dsp-eq-enable              (dc "rb_dsp_eq_enable" (fd-void [PTR BOOL]))
   :dsp-set-tone               (dc "rb_dsp_set_tone" (fd-void [PTR I32 I32]))
   :dsp-set-tone-cutoffs       (dc "rb_dsp_set_tone_cutoffs" (fd-void [PTR I32 I32]))
   :dsp-set-surround           (dc "rb_dsp_set_surround" (fd-void [PTR I32 I32 I32 I32]))
   :dsp-set-channel-config     (dc "rb_dsp_set_channel_config" (fd-void [PTR I32]))
   :dsp-set-stereo-width       (dc "rb_dsp_set_stereo_width" (fd-void [PTR I32]))
   :dsp-set-compressor         (dc "rb_dsp_set_compressor" (fd-void [PTR I32 I32 I32 I32 I32 I32]))
   :dsp-set-replaygain         (dc "rb_dsp_set_replaygain" (fd-void [PTR I32 BOOL F32]))
   :dsp-set-replaygain-gains   (dc "rb_dsp_set_replaygain_gains" (fd-void [PTR F32 F32 F32 F32]))
   :dsp-set-replaygain-gains-raw (dc "rb_dsp_set_replaygain_gains_raw" (fd-void [PTR I64 I64 I64 I64]))
   :dsp-set-eq-band            (dc "rb_dsp_set_eq_band" (fd-void [PTR I64 I32 F32 F32]))
   :dsp-set-eq-precut          (dc "rb_dsp_set_eq_precut" (fd-void [PTR F32]))
   :dsp-process                (dc "rb_dsp_process" (fd-of PTR [PTR PTR I64 PTR]))
   ;; ---- metadata ----
   :meta-read-json             (dc "rb_meta_read_json" (fd-of PTR [PTR]))
   :meta-probe                 (dc "rb_meta_probe" (fd-of PTR [PTR]))
   ;; ---- decoder ----
   :decoder-open               (dc "rb_decoder_open" (fd-of PTR [PTR]))
   :decoder-free               (dc "rb_decoder_free" (fd-void [PTR]))
   :decoder-metadata-json      (dc "rb_decoder_metadata_json" (fd-of PTR [PTR]))
   :decoder-next-chunk         (dc "rb_decoder_next_chunk" (fd-of PTR [PTR PTR PTR]))
   :decoder-seek-ms            (dc "rb_decoder_seek_ms" (fd-void [PTR I64]))
   :decoder-elapsed-ms         (dc "rb_decoder_elapsed_ms" (fd-of I64 [PTR]))
   :decoder-finished           (dc "rb_decoder_finished" (fd-of BOOL [PTR PTR]))
   ;; ---- player ----
   :player-new                 (dc "rb_player_new" (fd-of PTR []))
   :player-new-with-config     (dc "rb_player_new_with_config"
                                   (fd-of PTR [I32 F32 F32 I32 F32 BOOL I32 I32 I32 I32 I32 I32]))
   :player-new-with-config-ex  (dc "rb_player_new_with_config_ex"
                                   (fd-of PTR [I32 F32 F32 I32 F32 BOOL I32 I32 I32 I32 I32 I32 PTR I32]))
   :player-free                (dc "rb_player_free" (fd-void [PTR]))
   :player-set-queue-json      (dc "rb_player_set_queue_json" (fd-void [PTR PTR]))
   :player-enqueue             (dc "rb_player_enqueue" (fd-void [PTR PTR]))
   :player-insert-json         (dc "rb_player_insert_json" (fd-void [PTR PTR I32 I64]))
   :player-remove              (dc "rb_player_remove" (fd-void [PTR I64]))
   :player-clear-queue         (dc "rb_player_clear_queue" (fd-void [PTR]))
   :player-queue-json          (dc "rb_player_queue_json" (fd-of PTR [PTR]))
   :player-play                (dc "rb_player_play" (fd-void [PTR]))
   :player-pause               (dc "rb_player_pause" (fd-void [PTR]))
   :player-toggle              (dc "rb_player_toggle" (fd-void [PTR]))
   :player-stop                (dc "rb_player_stop" (fd-void [PTR]))
   :player-next                (dc "rb_player_next" (fd-void [PTR]))
   :player-previous            (dc "rb_player_previous" (fd-void [PTR]))
   :player-skip-to             (dc "rb_player_skip_to" (fd-void [PTR I64]))
   :player-seek-ms             (dc "rb_player_seek_ms" (fd-void [PTR I64]))
   :player-set-volume          (dc "rb_player_set_volume" (fd-void [PTR F32]))
   :player-set-balance         (dc "rb_player_set_balance" (fd-void [PTR I32]))
   :player-set-crossfade       (dc "rb_player_set_crossfade" (fd-void [PTR I32 I32 I32 I32 I32 I32]))
   :player-set-replaygain      (dc "rb_player_set_replaygain" (fd-void [PTR I32 F32 BOOL]))
   :player-set-shuffle         (dc "rb_player_set_shuffle" (fd-void [PTR BOOL]))
   :player-is-shuffle-enabled  (dc "rb_player_is_shuffle_enabled" (fd-of BOOL [PTR]))
   :player-set-repeat          (dc "rb_player_set_repeat" (fd-void [PTR I32]))
   :player-repeat              (dc "rb_player_repeat" (fd-of I32 [PTR]))
   ;; ---- player DSP ----
   :player-set-eq-enabled      (dc "rb_player_set_eq_enabled" (fd-void [PTR BOOL]))
   :player-is-eq-enabled       (dc "rb_player_is_eq_enabled" (fd-of BOOL [PTR]))
   :player-set-eq-band         (dc "rb_player_set_eq_band" (fd-void [PTR I64 I32 F32 F32]))
   :player-set-eq-precut       (dc "rb_player_set_eq_precut" (fd-void [PTR F32]))
   :player-set-eq-preset       (dc "rb_player_set_eq_preset" (fd-void [PTR I32]))
   :player-set-tone            (dc "rb_player_set_tone" (fd-void [PTR I32 I32 I32 I32]))
   :player-set-bass            (dc "rb_player_set_bass" (fd-void [PTR I32]))
   :player-set-treble          (dc "rb_player_set_treble" (fd-void [PTR I32]))
   :player-set-bass-cutoff     (dc "rb_player_set_bass_cutoff" (fd-void [PTR I32]))
   :player-set-treble-cutoff   (dc "rb_player_set_treble_cutoff" (fd-void [PTR I32]))
   :player-set-crossfeed       (dc "rb_player_set_crossfeed" (fd-void [PTR I32 I32 I32 I32 I32]))
   :player-set-bass-enhancement (dc "rb_player_set_bass_enhancement" (fd-void [PTR I32 I32]))
   :player-set-fatigue-reduction (dc "rb_player_set_fatigue_reduction" (fd-void [PTR I32]))
   :player-set-surround        (dc "rb_player_set_surround" (fd-void [PTR I32 I32 I32 I32]))
   :player-set-channel-mode    (dc "rb_player_set_channel_mode" (fd-void [PTR I32]))
   :player-set-stereo-width    (dc "rb_player_set_stereo_width" (fd-void [PTR I32]))
   :player-set-compressor      (dc "rb_player_set_compressor" (fd-void [PTR I32 I32 I32 I32 I32 I32]))
   :player-set-dither          (dc "rb_player_set_dither" (fd-void [PTR BOOL]))
   :player-set-pitch           (dc "rb_player_set_pitch" (fd-void [PTR I32]))
   :player-dsp-settings-json   (dc "rb_player_dsp_settings_json" (fd-of PTR [PTR]))
   :player-volume              (dc "rb_player_volume" (fd-of F32 [PTR]))
   :player-balance             (dc "rb_player_balance" (fd-of I32 [PTR]))
   :player-sample-rate         (dc "rb_player_sample_rate" (fd-of I32 [PTR]))
   :player-status-json         (dc "rb_player_status_json" (fd-of PTR [PTR]))
   ;; ---- resume ----
   :player-resume              (dc "rb_player_resume" (fd-of PTR [PTR]))
   :player-save-resume         (dc "rb_player_save_resume" (fd-void [PTR]))
   :player-clear-resume        (dc "rb_player_clear_resume" (fd-void [PTR]))
   :load-resume-json           (dc "rb_load_resume_json" (fd-of PTR [PTR]))
   ;; ---- m3u / m3u8 ----
   :player-import-m3u          (dc "rb_player_import_m3u" (fd-of PTR [PTR PTR I32 I64]))
   :player-load-m3u            (dc "rb_player_load_m3u" (fd-of PTR [PTR PTR]))
   :player-export-m3u          (dc "rb_player_export_m3u" (fd-of I32 [PTR PTR]))
   :m3u-read-json              (dc "rb_m3u_read_json" (fd-of PTR [PTR]))
   :m3u-write-json             (dc "rb_m3u_write_json" (fd-of I32 [PTR PTR]))
   :is-url                     (dc "rb_is_url" (fd-of BOOL [PTR]))}))

(defn call
  "Invoke the C function bound to `k`. Args must already be the right JVM type
  (use int/long/float/boolean/MemorySegment at the call site — the ABI does no
  narrowing). Returns the boxed result, or nil for void functions."
  [k & args]
  (.invokeWithArguments ^MethodHandle (get @handles k) (to-array args)))

(defn take-string
  "Copy a heap C string returned by the ABI into a String, then free it.
  Returns nil for a NULL pointer (the ABI's error/absent signal)."
  [^MemorySegment seg]
  (when (and seg (not (zero? (.address seg))))
    (let [s (.getString (.reinterpret seg Long/MAX_VALUE) 0)]
      (call :string-free seg)
      s)))

(defn abi-version
  "ABI major version of the loaded library (bumped on breaking changes)."
  []
  (bit-and (long (call :abi-version)) 0xFFFFFFFF))

File: bindings/ruby/lib/rockbox_ffi/ffi.rb
# frozen_string_literal: true

# Fiddle (Ruby stdlib) loader for librockbox_ffi.
#
# Declares exactly the functions we call and dlopens the prebuilt shared
# library — no C is compiled here. The library is located via the
# ROCKBOX_FFI_LIB env var, then by walking up to the repo's target/release
# directory. Mirrors include/rockbox_ffi.h — keep in sync with the C ABI.

require "fiddle"
require "fiddle/import"

module RockboxFFI
  # Locate the prebuilt shared library. Precedence:
  #   1. ROCKBOX_FFI_LIB env var (explicit override)
  #   2. the platform binary bundled in the gem's vendor/ dir (published gems)
  #   3. target/release, by walking up from here (repo checkout / dev)
  def self.library_path
    names = %w[librockbox_ffi.dylib librockbox_ffi.so rockbox_ffi.dll]
    tried = []

    if (env = ENV["ROCKBOX_FFI_LIB"])
      tried << env
      return env if File.exist?(env)
    end

    # Bundled binary (platform gems ship one under <gem_root>/vendor/).
    vendor = File.expand_path("../../vendor", __dir__)
    names.each do |name|
      cand = File.join(vendor, name)
      tried << cand
      return cand if File.exist?(cand)
    end

    # Walk up from this file looking for target/release (repo checkout).
    dir = File.expand_path(__dir__)
    loop do
      rel = File.join(dir, "target", "release")
      if File.directory?(rel)
        names.each do |name|
          cand = File.join(rel, name)
          tried << cand
          return cand if File.exist?(cand)
        end
      end
      parent = File.dirname(dir)
      break if parent == dir

      dir = parent
    end

    raise LoadError, "could not locate librockbox_ffi shared library. Set " \
                     "ROCKBOX_FFI_LIB or run `cargo build --release -p " \
                     "rockbox-ffi`. Tried:\n  #{tried.join("\n  ")}"
  end

  # Thin Fiddle wrappers over the C ABI. Every function is a module method
  # (e.g. `Lib.rb_dsp_new`). Opaque handles and char*/int16* returns come back
  # as Fiddle::Pointer; scalars come back as Integer/Float.
  module Lib
    extend Fiddle::Importer
    dlload RockboxFFI.library_path

    # Fixed-width C types → Fiddle's native aliases (64-bit LP64 assumed).
    typealias "uint32_t", "unsigned int"
    typealias "int32_t", "int"
    typealias "int16_t", "short"
    typealias "int64_t", "long long"
    typealias "uint64_t", "unsigned long long"
    typealias "size_t", "unsigned long"
    typealias "bool", "int" # passed as 0/1 in a full register on SysV/AAPCS

    extern "uint32_t rb_ffi_abi_version()"

    extern "void rb_string_free(void*)"
    extern "void rb_buffer_free(void*, size_t)"

    # ---- DSP ----------------------------------------------------------
    extern "void* rb_dsp_new(uint32_t)"
    extern "void rb_dsp_free(void*)"
    extern "void rb_dsp_set_input_frequency(void*, uint32_t)"
    extern "void rb_dsp_flush(void*)"
    extern "void rb_dsp_eq_enable(void*, bool)"
    extern "void rb_dsp_set_tone(void*, int32_t, int32_t)"
    extern "void rb_dsp_set_tone_cutoffs(void*, int32_t, int32_t)"
    extern "void rb_dsp_set_surround(void*, int32_t, int32_t, int32_t, int32_t)"
    extern "void rb_dsp_set_channel_config(void*, int32_t)"
    extern "void rb_dsp_set_stereo_width(void*, int32_t)"
    extern "void rb_dsp_set_compressor(void*, int32_t, int32_t, int32_t, int32_t, int32_t, int32_t)"
    extern "void rb_dsp_set_replaygain(void*, int32_t, bool, float)"
    extern "void rb_dsp_set_replaygain_gains(void*, float, float, float, float)"
    extern "void rb_dsp_set_replaygain_gains_raw(void*, int64_t, int64_t, int64_t, int64_t)"
    extern "void rb_dsp_set_eq_band(void*, size_t, int32_t, float, float)"
    extern "void rb_dsp_set_eq_precut(void*, float)"
    extern "void* rb_dsp_process(void*, void*, size_t, void*)"

    # ---- metadata -----------------------------------------------------
    extern "void* rb_meta_read_json(void*)"
    extern "void* rb_meta_probe(void*)"

    # ---- player -------------------------------------------------------
    extern "void* rb_player_new()"
    extern "void* rb_player_new_with_config(uint32_t, float, float, int32_t, float, bool, int32_t, uint32_t, uint32_t, uint32_t, uint32_t, int32_t)"
    extern "void* rb_player_new_with_config_ex(uint32_t, float, float, int32_t, float, bool, int32_t, uint32_t, uint32_t, uint32_t, uint32_t, int32_t, void*, uint32_t)"
    extern "void rb_player_free(void*)"
    extern "void rb_player_set_queue_json(void*, void*)"
    extern "void rb_player_enqueue(void*, void*)"
    extern "void rb_player_insert_json(void*, void*, int32_t, size_t)"
    extern "void* rb_player_queue_json(void*)"
    extern "void rb_player_play(void*)"
    extern "void rb_player_pause(void*)"
    extern "void rb_player_toggle(void*)"
    extern "void rb_player_stop(void*)"
    extern "void rb_player_next(void*)"
    extern "void rb_player_previous(void*)"
    extern "void rb_player_skip_to(void*, size_t)"
    extern "void rb_player_seek_ms(void*, uint64_t)"
    extern "void rb_player_set_volume(void*, float)"
    extern "void rb_player_set_crossfade(void*, int32_t, uint32_t, uint32_t, uint32_t, uint32_t, int32_t)"
    extern "void rb_player_set_replaygain(void*, int32_t, float, bool)"
    extern "float rb_player_volume(void*)"
    extern "uint32_t rb_player_sample_rate(void*)"
    extern "void* rb_player_status_json(void*)"

    # ---- player DSP ---------------------------------------------------
    extern "void rb_player_set_eq_enabled(void*, bool)"
    extern "bool rb_player_is_eq_enabled(void*)"
    extern "void rb_player_set_eq_band(void*, size_t, int32_t, float, float)"
    extern "void rb_player_set_eq_precut(void*, float)"
    extern "void rb_player_set_eq_preset(void*, int32_t)"
    extern "void rb_player_set_tone(void*, int32_t, int32_t, int32_t, int32_t)"
    extern "void rb_player_set_bass(void*, int32_t)"
    extern "void rb_player_set_treble(void*, int32_t)"
    extern "void rb_player_set_surround(void*, int32_t, int32_t, int32_t, int32_t)"
    extern "void rb_player_set_channel_mode(void*, int32_t)"
    extern "void rb_player_set_stereo_width(void*, int32_t)"
    extern "void rb_player_set_compressor(void*, int32_t, int32_t, int32_t, int32_t, int32_t, int32_t)"
    extern "void rb_player_set_dither(void*, bool)"
    extern "void rb_player_set_pitch(void*, int32_t)"
    extern "void* rb_player_dsp_settings_json(void*)"

    # ---- resume -------------------------------------------------------
    extern "void* rb_player_resume(void*)"
    extern "void rb_player_save_resume(void*)"
    extern "void rb_player_clear_resume(void*)"
    extern "void* rb_load_resume_json(void*)"

    # ---- m3u / m3u8 playlists -----------------------------------------
    extern "void* rb_player_import_m3u(void*, void*, int32_t, size_t)"
    extern "void* rb_player_load_m3u(void*, void*)"
    extern "int32_t rb_player_export_m3u(void*, void*)"
    extern "void* rb_m3u_read_json(void*)"
    extern "int32_t rb_m3u_write_json(void*, void*)"
    extern "bool rb_is_url(void*)"
  end

  # true/false → 1/0 for the ABI's `bool` (declared as int above).
  def self.b(flag)
    flag ? 1 : 0
  end

  # Copy a heap C string returned by the ABI into a String, then free it.
  # Returns nil for a NULL pointer (the ABI's error/absent signal).
  def self.take_string(ptr)
    return nil if ptr.null?

    begin
      ptr.to_s.force_encoding("UTF-8")
    ensure
      Lib.rb_string_free(ptr)
    end
  end
end
